addons/io/org.openhab.io.azureiothub/src/main/java/org/openhab/io/internal/azureiothub/AzureDevices.java

```diff
@@ -0,0 +1,78 @@
+/**
+ * Copyright (c) 2010-2018 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.io.internal.azureiothub;
+
+import java.time.Duration;
+import java.time.LocalDateTime;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.microsoft.azure.sdk.iot.service.Device;
+import com.microsoft.azure.sdk.iot.service.RegistryManager;
+import com.microsoft.azure.sdk.iot.service.exceptions.IotHubException;
+
+/**
+ * Represents an a list of devices on the Azure IoT Hub
+ *
+ * @author Niko Tanghe - Initial contribution
+ * @author Kai Kreuzer - removed joda-time dependency and cleaned up code
+ */
+
+public class AzureDevices {
+
+    private final Logger logger = LoggerFactory.getLogger(AzureDevices.class);
+
+    private Map<String, AzureDevice> map = new HashMap<String, AzureDevice>();
+    private String connectionstring;
+    private RegistryManager registryManager;
+
+    public AzureDevices(String connectionstring) throws Exception {
+        this.connectionstring = connectionstring;
+        this.registryManager = RegistryManager.createFromConnectionString(this.connectionstring);
+    }
+
+    public AzureDevice getDevice(String deviceId) throws Exception {
+        AzureDevice device = map.get(deviceId);
+
+        // not created yet or cache expired
+        if (device == null || device.getRetrievedAt().plus(Duration.ofMinutes(30)).isBefore(LocalDateTime.now())) {
+
+            // close connection
+            if (device != null) {
+                device.close();
+                registryManager.removeDevice(deviceId);
+            }
+
+            // re-create
+            device = new AzureDevice(getAzureDevice(deviceId), registryManager);
+            map.put(deviceId, device);
+            return device;
+        }
+
+        return device;
+    }
+
+    private Device getAzureDevice(String deviceId) throws Exception {
+        Device device = Device.createFromId(deviceId, null, null);
+
+        try {
+            device = registryManager.addDevice(device);
+        } catch (IotHubException iote) {
+            // this happens if the device already exists, so let's retrieve it
+            device = registryManager.getDevice(deviceId);
+        } catch (Exception ex) {
```

Comment:
Do you really need to catch the Exception-class or are there more specific exceptions you could catch like the IOException or maybe only RuntimeException's (which is not that good either). The current code and also catching RuntimeException's might 'hide' programming errors like NullPointerExceptions occurring.

https://www.google.nl/search?q=java%20why%20not%20to%20catch%20exception